pyquil/api/tests/test_quantum_computer.py

```diff
@@ -205,3 +205,30 @@ def test_qc_joint_calibration(forest):
     # Z1 expectation value for state |01> with 95% RO fid on both qubits is about 0.9
     assert np.isclose(results[0].additional_results[1].expectation, 0.9, atol=0.01)
     assert results[0].additional_results[1].total_counts == 40000
+
+
+def test_qc_expectation_on_qvm_that_requires_executable(forest):
+    # regression test for https://github.com/rigetti/forest-tutorials/issues/2
+    device = NxDevice(nx.complete_graph(2))
+    qc = QuantumComputer(
+        name="testy!",
+        qam=QVM(connection=forest, requires_executable=True),
+        device=device,
+        compiler=DummyCompiler(),
+    )
+
+    p = Program()
+    theta = p.declare("theta", "REAL")
+    p += RESET()
+    p += RX(theta, 0)
+    p.wrap_in_numshots_loop(2)
+
+    sx = ExperimentSetting(in_state=sZ(0) * sZ(1), out_operator=sX(0) * sX(1))
+    e = Experiment(settings=[sx], program=p)
+
+    thetas = np.linspace(start=-np.pi / 2, stop=np.pi / 2, num=3)
+
+    # Here we verify that multiple calls to qc.experiment with the same experiment backed by a QVM
+    # that requires_exectutable does not raise an exception.
+    for theta in thetas:
+        qc.experiment(e, memory_map={"theta": [theta]})
```

Comment:
No doubt this test could be improved to actually check the that results look correct, but then I'd have to understand what is going on 😬

Reply:
We can test it. For simplicity we could just make `in_state` and `out_state` single-qubit operators (i.e. get rid of `sZ(1)` and `sX(1)`).

This experiment says prepare my state using the Pauli operator `Z_0`, run the program `RX(theta) 0`, and then measure in the `X_0` basis. (Since we assume the hardware's default basis is `Z_0` the state prep part does nothing here.) You can probably convince yourself (by thinking about the Bloch sphere) that this example should give an expectation of close to zero for all `theta`.

For a less trivial example, you could change `RX(theta) 0` to `RY(theta) 0` -- that would give you (I think) expectation `-1` for `theta = -pi/2`, and `+1` for `theta = pi/2`, and something small for `theta = 0`.

Reply:
Thanks for the explanation (and for taking over ecp's unofficial role as quantum noob tutor :)

> We can test it. For simplicity we could just make `in_state` and `out_state` single-qubit operators (i.e. get rid of `sZ(1)` and `sX(1)`).

will do


> This experiment says prepare my state using the Pauli operator `Z_0`, run the program `RX(theta) 0`, and then measure in the `X_0` basis. (Since we assume the hardware's default basis is `Z_0` the state prep part does nothing here.) You can probably convince yourself (by thinking about the Bloch sphere) that this example should give an expectation of close to zero for all `theta`.

Assuming that measuring in the `X_0` basis is like projecting the state vector onto the x-axis, then yes I think I see what you mean.

I am making my way through Nielsen & Chuang at a snail's pace and it turns out my last stopping point was one page short of section 2.2.3 on Quantum Measurement, so maybe I will try to read that tonight (or at least the first two pages before I get exhausted and switch to cat videos). If I can make it as far as section 2.2.5 on Projective Measurement, then I'll have earned it.

> For a less trivial example, you could change `RX(theta) 0` to `RY(theta) 0` -- that would give you (I think) expectation `-1` for `theta = -pi/2`, and `+1` for `theta = pi/2`, and something small for `theta = 0`.

👍

Reply:
> Assuming that measuring in the X_0 basis is like projecting the state vector onto the x-axis, then yes I think I see what you mean.

Having read through the book sections mentioned above, I guess I'm either wrong about being able to interpret measurement in the X_0 basis as projection onto the x-axis in the bloch sphere, or else right by some coincidence I don't understand.

Anyway, I was able to churn through the equation given (2.110) for calculating the "average value" of a projective measurement. Is this the same as the expectation being calculated here? Seems plausible. At any rate, plugging in +1, `|+>` and -1, `|->` as the eigenvalues/eigenvectors of X and substituting the state `|psi> = RX (theta) 0` for the values of theta mentioned above does result in an average value of 0, so either I got the calculations wrong or that seems to check out. Likewise for the `RY (theta) 0` example you mentioned, with `|psi>` in `{|+>, |->, |0>}` giving the expectation values +/- 1 and 0, as you say.

What does it all mean? 🤷‍♂ 
What is the significance of the `_0` in "`X_0` basis"? 🤷‍♀ 
How does this calculation for the average value of a projective measurement correspond to the geometric interpretation on the bloch spehere that you hinted at above? 🤷‍♂ 🤷‍♀ 

I'm more confused now than I was before, but at least I have choco chip pancakes here to comfort me.

Reply:
> Since we assume the hardware's default basis is Z_0 the state prep part does nothing here.

More precisely, we choose to call the natural initial state of the hardware (which is the ground state of the transmon) the zero ket of the Z basis `|0>`.

Reply:
I agree with the statements regarding Rx(theta) and Ry(theta). If you like to think algebraically, you can see it this way: Applying ![R_x(\theta)](https://render.githubusercontent.com/render/math?math=R_x(%5Ctheta)) to ![\big| 0 \rangle](https://render.githubusercontent.com/render/math?math=%5Cbig%7C%200%20%5Crangle) and then measuring in the `X` basis gives an expectation of ![\langle 0 \big | R_x(-\theta) X R_x(\theta) \big | 0\rangle](https://render.githubusercontent.com/render/math?math=%5Clangle%200%20%5Cbig%20%7C%20R_x(-%5Ctheta)%20X%20R_x(%5Ctheta)%20%5Cbig%20%7C%200%5Crangle). The Taylor series for ![R_x(\theta)](https://render.githubusercontent.com/render/math?math=R_x(%5Ctheta)) involves only terms proportional to the identity and to `X`. These terms all commute with the factor of `X` in the expectation. The factors of + and - \theta, then cancel.

Reply:
@appleby, if I understand your reasoning, it looks correct. But, `X_0`-basis doesn't make sense. You are measuring in the `X` basis. There are a few ways to "explain" what this means (really just connect it to other stuff).  The unitary transform from the Z to X basis vectors is `H` (Its columns are the eigenvectors of X). And (noting that H' = H), you can transform the Z operator rather than the kets to take the expectation: `Z -> H Z H`. And `H Z H = X`. So measuring in the X basis, is equivalent to the expectation of `X` in the Z basis.
(There is no significant to the mishmash of fonts and quoting that I used)

Reply:
@jlapeyre thanks this is very helpful.

math noob question regarding your first comment: the fact that terms commute and cancel makes sense, but what does the Taylor series have to do with it?

Reply:
For example ![\exp(M) = I + M + M^2 / 2 + \ldots](https://render.githubusercontent.com/render/math?math=%5Cexp(M)%20%3D%20I%20%2B%20M%20%2B%20M%5E2%20%2F%202%20%2B%20%5Cldots).  So, if ![A](https://render.githubusercontent.com/render/math?math=A) commutes with ![M](https://render.githubusercontent.com/render/math?math=M), then it commutes with ![\exp(M)](https://render.githubusercontent.com/render/math?math=%5Cexp(M)). This is because ![A](https://render.githubusercontent.com/render/math?math=A) commutes with any power of ![A](https://render.githubusercontent.com/render/math?math=M), and so, with each term in the series.

So, in the original example, you commute X through one of the rotation operators. Then the two rotation operators are adjacent, so they cancel.

Reply:
A simpler way to see this is to note that X is proportional to RX(pi). Then the term in your expectation becomes RX(-theta) RX(pi) RX(theta) = RX(-theta + pi + theta) = RX(pi) = X (up to complex scalars with unit modulus).

In general, to see that RX(alpha) commutes with RX(beta) just think that rotating something along an axis V by an angle alpha and then rotating the result along the same axis by an angle beta is equivalent to rotating by beta first and then rotating by alpha (thus, you get the commutation relation [RX(alpha), RX(beta)] = 0).